pageserver/src/tenant.rs

```diff
@@ -1244,32 +1249,50 @@ impl Tenant {
     }
 
     /// Removes timeline-related in-memory data
-    pub fn delete_timeline(&self, timeline_id: TimelineId) -> anyhow::Result<()> {
-        // in order to be retriable detach needs to be idempotent
-        // (or at least to a point that each time the detach is called it can make progress)
-        let mut timelines = self.timelines.lock().unwrap();
+    pub async fn delete_timeline(&self, timeline_id: TimelineId) -> anyhow::Result<()> {
+        // Transition the timeline into TimelineState::Paused
+        let timeline = {
+            let mut timelines = self.timelines.lock().unwrap();
 
-        // Ensure that there are no child timelines **attached to that pageserver**,
-        // because detach removes files, which will break child branches
-        let children_exist = timelines
-            .iter()
-            .any(|(_, entry)| entry.get_ancestor_timeline_id() == Some(timeline_id));
+            // Ensure that there are no child timelines **attached to that pageserver**,
+            // because detach removes files, which will break child branches
```

Comment:
What happens if this is called concurrently from two threads? Should we check here that the timeline is not already in Paused state?

Reply:
Oh I see there's some code to deal with that race later, when the Timeline is removed from the map. I think you can get errors from `std::fs::remove_dir_all`, however, if two threads try to remove the files at the same time.

Reply:
Yeah, one will fail with an error and another will succeed.
Since we don't (and can't!) rollback the state from `Paused` to `Active`, I think that's acceptable for now.

Reply:
I'm worried that they would both fail. Imagine that you have a directory with two files, "foo" and "bar". One thread removes "foo" first, then "bar", and the other one the other way round. Both could succeed on removing the first file, and get a "file not found" error on the second one. I haven't looked at how `std::fs::remove_dir_all` is implemented, so I'm not sure what would happen in that case, but I also don't see any guarantees in the docs for that either.